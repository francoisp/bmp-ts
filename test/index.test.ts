import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';

import bmp from '../lib/';

const createPath = (p: string) => path.join(process.cwd(), p);

function checksum(str: string, algorithm = 'md5', encoding = 'hex') {
  return (
    crypto
      .createHash(algorithm)
      .update(str, 'utf8')
      // @ts-ignore
      .digest(encoding)
  );
}

describe('BMP', () => {
  describe('decode', () => {
    test('errors for non bmp files', () => {
      const buff = fs.readFileSync(createPath('package.json'));
      expect(() => bmp.decode(buff)).toThrow();
    });

    test('1-bit', () => {
      const buff = fs.readFileSync(createPath('old-tests/bit1.bmp'));
      const res = checksum(bmp.decode(buff).data.toString());
      expect(res).toMatchSnapshot();
    });

    describe('4-bit', () => {
      test('normal', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit4.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('RLE', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit4_RLE.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });
    });

    describe('8-bit', () => {
      test('normal', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit8.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('RLE', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit8_RLE.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });
    });

    describe('16-bit', () => {
      test('555', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit16_x555.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('444', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit16_x444.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('555', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit16_x555.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('444', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit16_a444.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('555', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit16_a555.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('565', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit16_565.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });
    });

    test('24-bit', () => {
      const buff = fs.readFileSync(createPath('old-tests/bit24.bmp'));
      const res = checksum(bmp.decode(buff).data.toString());
      expect(res).toMatchSnapshot();
    });

    describe('32-bit', () => {
      test('normal', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit32.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('alpha', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit32_alpha.bmp'));
        const res = checksum(bmp.decode(buff).data.toString());
        expect(res).toMatchSnapshot();
      });

      test('alpha - toRGBA', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit32_alpha.bmp'));
        const res = checksum(
          bmp.decode(buff, { toRGBA: true }).data.toString()
        );
        expect(res).toMatchSnapshot();
      });
    });
  });

  describe('encode', () => {
    test('1-bit', () => {
      const buff = fs.readFileSync(createPath('old-tests/bit32.bmp'));
      const bitmap = bmp.decode(buff);

      bitmap.bitPP = 1;

      const res = checksum(bmp.encode(bitmap).data.toString());
      expect(res).toMatchSnapshot();
    });

    describe('4-bit', () => {
      test('errors without color pallette', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit32.bmp'));
        const bitmap = bmp.decode(buff);

        bitmap.bitPP = 4;

        expect(() => bmp.encode(bitmap)).toThrow();
      });

      test('works when colors provided', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit4.bmp'));
        const bitmap = bmp.decode(buff);

        bitmap.bitPP = 4;

        const res = checksum(bmp.encode(bitmap).data.toString());
        expect(res).toMatchSnapshot();
      });
    });

    describe('8-bit', () => {
      test('errors without color pallette', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit32.bmp'));
        const bitmap = bmp.decode(buff);

        bitmap.bitPP = 8;

        expect(() => bmp.encode(bitmap)).toThrow();
      });

      test('works when colors provided', () => {
        const buff = fs.readFileSync(createPath('old-tests/bit8.bmp'));
        const bitmap = bmp.decode(buff);

        bitmap.bitPP = 8;

        const res = checksum(bmp.encode(bitmap).data.toString());
        expect(res).toMatchSnapshot();
      });
    });

    test('16-bit', () => {
      const buff = fs.readFileSync(createPath('old-tests/bit32.bmp'));
      const bitmap = bmp.decode(buff);

      bitmap.bitPP = 16;

      const res = checksum(bmp.encode(bitmap).data.toString());
      expect(res).toMatchSnapshot();
    });

    test('24-bit', () => {
      const buff = fs.readFileSync(createPath('old-tests/bit32.bmp'));
      const bitmap = bmp.decode(buff);

      bitmap.bitPP = 24;

      const res = checksum(bmp.encode(bitmap).data.toString());
      expect(res).toMatchSnapshot();
    });

    test('32-bit', () => {
      const buff = fs.readFileSync(createPath('old-tests/bit32.bmp'));
      const bitmap = bmp.decode(buff);

      bitmap.bitPP = 32;

      const res = checksum(bmp.encode(bitmap).data.toString());
      expect(res).toMatchSnapshot();
    });
  });
});
