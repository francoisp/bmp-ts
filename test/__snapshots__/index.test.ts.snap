// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BMP decode 1-bit 1`] = `"cec76def32993b226b554d08f78948d9"`;

exports[`BMP decode 4-bit RLE 1`] = `"beac91cf64120346736c9b1383f9da60"`;

exports[`BMP decode 4-bit normal 1`] = `"df219955a66ea20a8ccb4f0537930458"`;

exports[`BMP decode 8-bit RLE 1`] = `"81bbf4f1f828dde5665bf1d798e39e72"`;

exports[`BMP decode 8-bit normal 1`] = `"6697bb6ab8078d739fc40bbe558bf4ce"`;

exports[`BMP decode 16-bit 444 1`] = `"988a4d08bdaf12a32954e81f5794503a"`;

exports[`BMP decode 16-bit 444 2`] = `"e16a3905a7a21a0e8f504285fccb3544"`;

exports[`BMP decode 16-bit 555 1`] = `"fea6f96b3099726499275ae8b395df02"`;

exports[`BMP decode 16-bit 555 2`] = `"fea6f96b3099726499275ae8b395df02"`;

exports[`BMP decode 16-bit 555 3`] = `"6fb3e1971fa948d1cc1fe268a037f05a"`;

exports[`BMP decode 16-bit 565 1`] = `"aa2751f1e5c305448a9622a926ca3628"`;

exports[`BMP decode 24-bit 1`] = `"389fa2b1873b5aa14f7ecab58fe96bd9"`;

exports[`BMP decode 32-bit alpha - toRGBA 1`] = `"1fb6cd994623e11eff63c44693a405ad"`;

exports[`BMP decode 32-bit alpha 1`] = `"5cb1c7699a421a7bead9adc527e9da8f"`;

exports[`BMP decode 32-bit normal 1`] = `"5cb1c7699a421a7bead9adc527e9da8f"`;

exports[`BMP encode 1-bit 1`] = `"078d88f8242ceacbc12ced3af9ed5c03"`;

exports[`BMP encode 4-bit works when colors provided 1`] = `"bfae56540db1291326a6f9f62cf6f35b"`;

exports[`BMP encode 8-bit works when colors provided 1`] = `"1e911ae143c47c2264f66b8f40a32e5d"`;

exports[`BMP encode 16-bit 1`] = `"649365cfd6af25db632ade4ed84345d4"`;

exports[`BMP encode 24-bit 1`] = `"6001e6e75e7dea586aeb617a5c8026fb"`;

exports[`BMP encode 32-bit 1`] = `"6bc4f43b00e2d88e2c553278472243b7"`;
